modules/network/src/main/java/org/apache/ignite/network/MessagingService.java

```diff
@@ -61,12 +61,23 @@
      * returns a future that will be completed successfully upon receiving a response.
      *
      * @param recipient Recipient of the message.
-     * @param msg A message.
+     * @param msg The message.
      * @param timeout Waiting for response timeout in milliseconds.
      * @return A future holding the response or error if the expected response was not received.
      */
     CompletableFuture<NetworkMessage> invoke(ClusterNode recipient, NetworkMessage msg, long timeout);
 
+    /**
+     * Sends a message asynchronously with same guarantees as {@link #send(ClusterNode, NetworkMessage)} and
+     * returns a future that will be completed successfully upon receiving a response.
+     *
+     * @param addr Recipient network address in host:port format.
+     * @param msg A message.
+     * @param timeout Waiting for response timeout in milliseconds.
+     * @return A future holding the response or error if the expected response was not received.
+     */
+    CompletableFuture<NetworkMessage> invoke(String addr, NetworkMessage msg, long timeout);
```

Comment:
Why we need it?

Reply:
To clearly separate messaging from discovery.
You don't need a node to be discovered to send a message to it and it's better fits to raft communication model, code is much cleaner. Moreover, ClusterNode has the address [1]
[1] org.apache.ignite.network.ClusterNode#address